src/shared/tools.ts

```diff
@@ -267,3 +269,85 @@ export interface DiffStrategy {
 
 	getProgressStatus?(toolUse: ToolUse, result?: any): ToolProgressStatus
 }
+
+/**
+ * Safely reads a file from the .roo directory with the given filename.
+ * Returns an empty string if the file doesn't exist or cannot be read.
+ *
+ * @param cwd - The working directory
+ * @param file - The filename to read from the .roo directory
+ * @returns A promise that resolves to the file content or empty string
+ */
+export async function readToolOverride(cwd: string, file: string): Promise<string> {
+	try {
+		const filePath = path.join(cwd, ".roo", "tools", `${file}.md`)
+		const content = await fs.readFile(filePath, "utf-8")
+		return content.trim()
+	} catch (err) {
+		const errorCode = (err as NodeJS.ErrnoException).code
+		if (!errorCode || !["ENOENT", "EISDIR"].includes(errorCode)) {
+			throw err
+		}
+		return ""
+	}
+}
+
+/**
+ * Interpolates args data into a tool description string.
+ * Replaces placeholders like ${args.cwd} or ${args.cwd.toPosix()} with actual values from the args object.
+ *
+ * @param content - The tool description content with placeholders
+ * @param args - The tool arguments object containing values to interpolate
+ * @returns The content with interpolated values
+ */
+function interpolateToolDescription(content: string, args: any): string {
+	// Replace ${args.property.method()} or ${args.property} patterns with actual values
+	return content.replace(/\$\{args\.([^}]+)\}/g, (match, expression) => {
+		try {
```

Comment:
make sure you support escaping symbols so that users can embed `${ foo }` without breaking things as `\${ foo }`

see above for an example with a negative look ahead.